open_spiel/games/battleship.h

```diff
@@ -0,0 +1,365 @@
+class BattleshipState final : public State {
+ public:
+  explicit BattleshipState(const std::shared_ptr<const BattleshipGame> bs_game);
+  ~BattleshipState() = default;
+
+  // Virtual functions inherited by OpenSpiel's `State` interface
+  Player CurrentPlayer() const override;
+  std::vector<Action> LegalActions() const override;
+  std::string ActionToString(Player player, Action action) const override;
+  std::string ToString() const override;
+  bool IsTerminal() const override;
+  std::vector<double> Returns() const override;
+  std::unique_ptr<State> Clone() const override;
+  std::string InformationStateString(Player player) const override;
+  std::string ObservationString(Player player) const override;
+  void UndoAction(Player player, Action action) override;
+
+  // Draws the board of a player.
+  //
+  // The board is drawn as a rectangular grid of characters, with the following
+  // conventions:
+  // - Ships are identified with letters, starting from 'a'. For each ship, we
+  //   mark all of the cells it occupies with the same letter.
+  //   - Lowercase letters denote that the cell was never hit by the opponent.
+  //   - Uppercase letters denote that the cell was hit by the opponent.
+  // - Cells marked with '*' denote shots by the opponent, that hit water.
+  // - All other cells are empty, that is, filled with a space ' ' character.
+  //
+  //
+  // Example
+  // -------
+  //
+  // This is what a typical 3x5 board string might looks like after 4 shots by
+  // the opponent.
+  //
+  // ```
+  // +------+
+  // |*a    |
+  // | A*   |
+  // |   bbB|
+  // +------+
+  // ```
+  std::string OwnBoardString(const Player player) const;
+
+  // Draws the state of the player's shots for far.
+  // This corresponds to the incremental board that the player builds over time
+  // by shooting at the opponent.
+  //
+  // The board is drawn as a rectangular grid of characters, with the following
+  // conventions:
+  // - Shots that hit a ship are marked with '#'.
+  // - Shots that hit the water are marked with '@'.
+  // - All other cells are empty, that is, filled with a space ' ' character.
+  //
+  //
+  // Example
+  // -------
+  //
+  // This is what the opponent player to the example provided in
+  // `OwnBoardString` will see:
+  //
+  // ```
+  // +------+
+  // |@     |
+  // | #@   |
+  // |     #|
+  // +------+
+  // ```
+  std::string ShotsBoardString(const Player player) const;
+
+  // OUtputs a pretty representation of the boards of the two players. This can
+  // be used as a more expensive (but readable) alternative to `ToString`.
+  //
+  // FIXME(gfarina): Decide is this should take the place of `ToString`.
+  //     @michalsustr suggested that would be a good idea. The only worry is
+  //     that it would create way longer state strings, and if implemented
+  //     algorithms keep internal maps keyed on state strings, that could lead
+  //     to performance issues.
+  std::string ToPrettyString() const;
```

Comment:
Ah, good to know. I will have ToPrettyString become ToString then, and remove the more "compact" representation of state.

Just to be sure: currently ToString captures more information that ToPrettyString, because it also contains information about the order of moves. So now we could have different histories be mapped to the same ToPrettyString, is that OK?

Reply:
> So now we could have different histories be mapped to the same ToPrettyString, is that OK?

Yes that it is fine, it's the same like permuting the order of moves in go.

Reply:
Perfect, got it. Thanks for confirming :-)